miden-tx/tests/test_miden_faucet_contract.rs

```diff
@@ -0,0 +1,433 @@
+};
+
+use assembly::ast::{ModuleAst, ProgramAst};
+use common::NodeIndex;
+use crypto::{hash::Hasher, Felt, StarkField, ZERO};
+
+use miden_objects::{
+    accounts::{Account, AccountCode, AccountId, AccountVault},
+    assets::{Asset, FungibleAsset},
+    notes::{Note, NoteEnvelope, NoteMetadata, NoteOrigin, NoteStub, NoteVault},
+    BlockHeader, ChainMmr,
+};
+use mock::{
+    account::{mock_account_storage, MockAccountType},
+    constants::ACCOUNT_ID_FUNGIBLE_FAUCET_ON_CHAIN,
+    notes::AssetPreservationStatus,
+    transaction::mock_inputs_with_existing,
+    utils::prepare_word,
+};
+
+use miden_tx::{DataStore, TransactionExecutor};
+use vm_processor::{crypto::Rpo256, Digest, MemAdviceProvider};
+
+#[derive(Clone)]
+pub enum DataStoreError {
+    AccountNotFound(AccountId),
+    NoteNotFound(u32, NodeIndex),
+}
+
+#[derive(Clone)]
+
+pub struct MockDataStore {
+    pub account: Account,
+    pub block_header: BlockHeader,
+    pub block_chain: ChainMmr,
+    pub notes: Vec<Note>,
+}
+
+impl MockDataStore {
+    pub fn with_existing(account: Option<Account>, consumed_notes: Option<Vec<Note>>) -> Self {
+        let (account, block_header, block_chain, consumed_notes) = mock_inputs_with_existing(
+            MockAccountType::StandardExisting,
+            AssetPreservationStatus::Preserved,
+            account,
+            consumed_notes,
+        );
+        Self {
+            account,
+            block_header,
+            block_chain,
+            notes: consumed_notes,
+        }
+    }
+}
+
+impl DataStore for MockDataStore {
+    fn get_transaction_data(
+        &self,
+        account_id: AccountId,
+        block_num: u32,
+        notes: &[NoteOrigin],
+    ) -> Result<(Account, BlockHeader, ChainMmr, Vec<Note>), miden_tx::DataStoreError> {
+        assert_eq!(account_id, self.account.id());
+        assert_eq!(block_num as u64, self.block_header.block_num().as_int());
+        assert_eq!(notes.len(), self.notes.len());
+        let origins = self
+            .notes
+            .iter()
+            .map(|note| note.proof().as_ref().unwrap().origin())
+            .collect::<Vec<_>>();
+        notes.iter().all(|note| origins.contains(&note));
+        Ok((
+            self.account.clone(),
+            self.block_header.clone(),
+            self.block_chain.clone(),
+            self.notes.clone(),
+        ))
+    }
+
+    fn get_account_code(
+        &self,
+        account_id: AccountId,
+    ) -> Result<ModuleAst, miden_tx::DataStoreError> {
+        assert_eq!(account_id, self.account.id());
+        Ok(self.account.code().module().clone())
+    }
+}
+
+fn get_faucet_account_with_max_suppply(max_supply: u64) -> Account {
+    let faucet_account_id = AccountId::try_from(ACCOUNT_ID_FUNGIBLE_FAUCET_ON_CHAIN).unwrap();
+    let faucet_account_code_src = include_str!("../../miden-lib/asm/faucets/basic.masm");
+    let faucet_account_code_ast = ModuleAst::parse(faucet_account_code_src).unwrap();
```

Comment:
Ideally, we should note use the path here to get the module AST. A better option would be to get it directly from `MidenLib`. We can do this right now by iterating over modules and picking the one with path `faucets/basic.masm`. But a better option would be to add `get_module(path)` metho to the `Library` trait. Let's create an issue for this in Miden VM repo.